subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/recomp/JavaRecompilationSpecProvider.java

```diff
@@ -106,12 +113,49 @@ public WorkResult decorateResult(RecompilationSpec recompilationSpec, WorkResult
     }
 
     private void processOtherChanges(CurrentCompilation current, PreviousCompilation previous, RecompilationSpec spec) {
+        if (spec.isFullRebuildNeeded()) {
+            return;
+        }
+        boolean emptyAnnotationProcessorPath = current.getAnnotationProcessorPath().isEmpty();
         SourceFileChangeProcessor javaChangeProcessor = new SourceFileChangeProcessor(previous);
-        AnnotationProcessorChangeProcessor annotationProcessorChangeProcessor = new AnnotationProcessorChangeProcessor(current, previous);
-        ResourceChangeProcessor resourceChangeProcessor = new ResourceChangeProcessor(current.getAnnotationProcessorPath());
-        InputChangeAction action = new InputChangeAction(spec, javaChangeProcessor, annotationProcessorChangeProcessor, resourceChangeProcessor, sourceFileClassNameConverter);
-        inputs.outOfDate(action);
-        inputs.removed(action);
+        for (FileChange fileChange : sourceFileChanges.get()) {
+            File file = fileChange.getFile();
+            if (fileChange.getFileType() != FileType.FILE) {
+                continue;
+            }
+            if (hasExtension(file, ".java")) {
+                Collection<String> classNames = sourceFileClassNameConverter.getClassNames(file);
+                if (classNames.isEmpty()) {
+                    // https://github.com/gradle/gradle/issues/9380
+                    // Remove a srcDir from a sourceSet
+                    spec.setFullRebuildCause("source dirs are changed", file);
+                    return;
+                } else {
+                    javaChangeProcessor.processChange(file, classNames, spec);
+                }
+            } else {
+                if (emptyAnnotationProcessorPath) {
+                    continue;
+                }
+                String changeName = determineChangeName(fileChange.getChangeType());
+                spec.setFullRebuildCause(fileChange.getFile().getName() + " has been " + changeName, null);
+                return;
+
+            }
+        }
+    }
+
+    private String determineChangeName(ChangeType changeType) {
```

Comment:
You could use an immutable enum map instead of you don't want to simply do `name().toLowerCase()`. Another option is to add `getDisplayName()` to `ChangeType`.

Reply:
I am using `name().toLowerCase()` now.